File: isSymmetric.java
import java.util.*;
// Approach 1- Recursive (DFS) [using implicit call stack]
class Solution{ // TC O(n) SC O(n)
    public boolean isSymmetric(Node root) {
    return isMirror(root, root);
}

public boolean isMirror(Node t1, Node t2) {
    if (t1 == null && t2 == null) return true;
    if (t1 == null || t2 == null) return false;
    return (t1.val == t2.val)
        && isMirror(t1.right, t2.left)
        && isMirror(t1.left, t2.right);
    }
}

// Approach 2- Iterative (BFS) [using queue]
class Solution1{
    public boolean isSymmetric(Node root) {
        Queue<Node> queue=new LinkedList<>();
        queue.add(root);
        queue.add(root);
        while(!queue.isEmpty()){
            Node t1=queue.poll();
            Node t2=queue.poll();
            if(t1==null && t2==null) continue;
            if(t1==null || t2==null) return false;
            if(t1.val!=t2.val) return false;
            queue.add(t1.left);
            queue.add(t2.right);
            queue.add(t1.right);
            queue.add(t2.left);
        }
        return true;
    }
}

class isSymmetric{
    public static void main(String[] args) {
        BinaryTree bt = new BinaryTree();
        bt.root=new Node(1);
        bt.root.left=new Node(2);
        bt.root.right=new Node(2);
        bt.root.left.left=new Node(3);
        bt.root.left.right=new Node(4);
        bt.root.right.left=new Node(4);
        bt.root.right.right=new Node(3);
        Solution s= new Solution();
        System.out.println(s.isSymmetric(bt.root));
        Solution1 s1= new Solution1();
        System.out.println(s1.isSymmetric(bt.root));
    }
}

File: isSymmetric.py
from collections import deque
class TreeNode:
    def __init__(self, val=None, left=None, right=None):
        self.val=val 
        self.left=left 
        self.right=right 

class Solution:
# Approach 1- Recursive (DFS) [using implicit call stack]
    def isSymmetric(self, root): # TC O(n) // SC O(n)

        def isMirror(t1,t2):
            if not t1 and not t2: return True 
            if not t1 or not t2: return False 
            return t1.val==t2.val and isMirror(t1.left, t2.right) and isMirror(t1.right, t2.left)

        return isMirror(root,root)

# Approach 2- Iterative (BFS) [using queue]
    def isSymmetric1(self, root): # TC O(n) // SC O(n)
        q=deque()
        q.append(root)
        q.append(root)
        while q:
            t1,t2=q.popleft(), q.popleft()
            if not t1 and not t2: continue 
            if not t1 or not t2: return False 
            if t1.val!=t2.val: return False 
            q.append(t1.left)
            q.append(t2.right)
            q.append(t1.right)
            q.append(t2.left)
        
        return True

s=Solution()
root=TreeNode(1)
root.left=TreeNode(2)
root.right=TreeNode(2)
root.left.left=TreeNode(4)
root.left.right=TreeNode(3)
root.right.left=TreeNode(3)
root.right.right=TreeNode(4)
print(s.isSymmetric(root))
print(s.isSymmetric1(root))

